src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -271,30 +284,112 @@ type Levels m h = V.Vector (Level m h)
-    -- | A merging of multiple runs.
-    MergingRun
-      !MergePolicyForLevel
-      !NumRuns
-      !NumEntries
-      -- ^ Sum of number of entries in the input runs
-      !(UnspentCreditsVar (PrimState m))
-      -- ^ The number of currently /unspent/ credits
-      !(TotalStepsVar (PrimState m))
-      -- ^ The total number of performed merging steps.
-      !(MutVar (PrimState m) MergeKnownCompleted)
-      -- ^ A variable that caches knowledge about whether the merge has been
-      -- completed. If 'MergeKnownCompleted', then we are sure the merge has
-      -- been completed, otherwise if 'MergeMaybeCompleted' we have to check the
+-- | An incoming run is either a single run, or a merge.
+data IncomingRun m h =
+    Single !(Run m h)
+  | Merging !(MergingRun m h)
+
+-- | A merging of multiple runs.
+-- TODO: Move to a separate module.
+data MergingRun m h = MergingRun {
+      mergePolicy         :: !MergePolicyForLevel
+    , mergeNumRuns        :: !NumRuns
+      -- | Sum of number of entries in the input runs
+    , mergeNumEntries     :: !NumEntries
+      -- | The number of currently /unspent/ credits
+    , mergeUnspentCredits :: !(UnspentCreditsVar (PrimState m))
+      -- | The total number of performed merging steps.
+    , mergeStepsPerformed :: !(TotalStepsVar (PrimState m))
+      -- | A variable that caches knowledge about whether the merge has been
+      -- completed. If 'MergeKnownCompleted', then we are sure the merge has been
+      -- completed, otherwise if 'MergeMaybeCompleted' we have to check the
       -- 'MergingRunState'.
-      !(StrictMVar m (MergingRunState m h))
-    -- | The result of merging a single run, is a single run.
-  | SingleRun !(Run m h)
+    , mergeKnownCompleted :: !(MutVar (PrimState m) MergeKnownCompleted)
+    , mergeState          :: !(StrictMVar m (MergingRunState m h))
+    , mergeRefCounter     :: !(RC.RefCounter m)
+    }
+
+{-# SPECIALISE newMergingRun ::
+     MergePolicyForLevel
+  -> NumRuns
+  -> NumEntries
+  -> MergeKnownCompleted
+  -> MergingRunState IO h
+  -> IO (MergingRun IO h)
+  #-}
+-- This allows constructing ill-formed MergingRuns, but the flexibility is
+-- needed for creating a merging run that is already Completed, as well as
+-- opening a merging run from a snapshot.
+-- TODO: instead create a Single run when OneShot merging?
+-- TODO: do not store MergeKnownCompleted in snapshot? It's redundant.
+-- TODO: slightly different API for opening from snapshot?
+newMergingRun ::
+     (MonadMVar m, MonadMask m, MonadSTM m, MonadST m)
+  => MergePolicyForLevel
+  -> NumRuns
+  -> NumEntries
+  -> MergeKnownCompleted
+  -> MergingRunState m h
+  -> m (MergingRun m h)
+newMergingRun mergePolicy mergeNumRuns mergeNumEntries knownCompleted state = do
+    mergeUnspentCredits <- UnspentCreditsVar <$> newPrimVar 0
+    mergeStepsPerformed <- TotalStepsVar <$> newPrimVar 0
+    case state of
+      OngoingMerge{}   -> assert (knownCompleted == MergeMaybeCompleted) (pure ())
```

Comment:
`newMergingRun` is used rarely enough that it can be a proper error instead of an assertion, IMO

Reply:
I hope to remove this code soon either way, as I don't think one should be able to pass in an invalid `MergeKnownCompleted` in the first place.